src/java/org/apache/cassandra/db/compaction/BackgroundCompactionRunner.java

```diff
@@ -188,34 +188,36 @@ public void run()
                          cfs.name,
                          cfs.getCompactionStrategy().getName());
 
-            Supplier<CompletableFuture<?>> compactionTask = Optional.ofNullable(lazyRunCompactionTasks(cfs))
-                                                                    .orElseGet(() -> lazyRunUpgradeTasks(cfs));
+            Supplier<CompletableFuture<?>[]> compactionTasks = lazyRunCompactionTasks(cfs);
+            if (compactionTasks == null)
+                compactionTasks = lazyRunUpgradeTasks(cfs);
 
-            if (compactionTask != null)
+            if (compactionTasks != null)
             {
                 ongoingCompactions.add(cfs);
 
-                compactionTask.get().handle((ignored, throwable) -> {
+                CompletableFuture<?>[] tasksResults = compactionTasks.get();
```

Comment:
Could we add some explanation that the supplier is used to guarantee `ongoingCompactions` is set before the tasks are started?